### PR TITLE
Apps 5750 add testrail api and git commands

### PR DESCRIPTION
**Pull request by refael-m**
[APPS-5750](https://cloudinary.atlassian.net/browse/APPS-5750)  
wdio-allure-ts side

### CODE REVIEW

File: src/commons/TestRailApi.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    return this.instance;
+  }
+
+  /**
+   *   Change test 'Automation' field
+   *   @param testID test id as appear on testRail site
+   *   @param value choose one of AutomationOptions
+   */
+  public changeTestAutomationField(testID: string, value: AutomationFieldOptions): AxiosPromise {
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:34]: it need to be more generic like "updateFieldInTestRails" and it should get
`testId:string, {key, value}`

File: src/commons/GitHooks.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * Git hooks scripts
+ */
+export namespace GitHooks {
+  /**
+   * Update automation field on testrail for the last merge tests files
+   */
+  export async function setLastMergedTestsAsAutomatedInTestrail(): Promise<void> {
+    const filesList: string[] = GitUtils.getLastMergedFiles();
+
+    for (const file of filesList) {
```
Comment (FelixZilber) [inline src/commons/GitHooks.ts:16]: move it to separate function

File: src/commons/GitUtils.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { execSync as execAsync } from 'child_process';
```
Comment (FelixZilber) [inline src/commons/GitUtils.ts:1]: please update

File: src/commons/TestRailApi.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  }
+
+  /**
+   * Standard api request
+   * @param reqMethod request method
+   * @param reqUrl request url
+   * @param reqData json data
+   */
+  private makeRequest(reqMethod: string, reqUrl: string, reqData?: object): AxiosPromise {
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:57]: reqMethod is not a string

File: src/commons/TestRailApi.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    this.initAuth();
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:58]: consider move it to the constructor instead

**Review by FelixZilber — CHANGES_REQUESTED** [review]

File: src/commons/GitUtils.ts (hunk 3 above)
Comment (refael-m) [reply src/commons/GitUtils.ts:1]: done

File: src/commons/GitHooks.ts (hunk 2 above)
Comment (refael-m) [reply src/commons/GitHooks.ts:16]: done

File: src/commons/TestRailApi.ts (hunk 5 above)
Comment (refael-m) [reply src/commons/TestRailApi.ts:58]: done

File: src/commons/GitUtils.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+const CURRENT_BRANCH_NAME_SCRIPT: string = 'git rev-parse --abbrev-ref HEAD';
+
+/**
+ * Manage git commands
+ */
+export namespace GitUtils {
+  /**
+   * Return list of all files from the last merge.
+   */
+  export function getLastMergedFiles(): string[] {
```
Comment (FelixZilber) [inline src/commons/GitUtils.ts:16]: should it be exprted ?

File: src/commons/GitUtils.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  export function getLastMergedTestsIds(): string[] {
+    const mergedTestsFiles: string[] = getTestsFiles(getLastMergedFiles());
+    return extractTestIdFromFiles(mergedTestsFiles);
+  }
+
+  /**
+   * Return the name of the current branch
+   */
+  export function getCurrentBranchName(): string {
```
Comment (FelixZilber) [inline src/commons/GitUtils.ts:32]: is it needed?

File: src/commons/GitUtils.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  return fileName.endsWith(shouldEndWith) && new RegExp(shouldStartWithRegex).test(fileName);
+}
+
+/**
+ * Extract testIds from test files
+ * @param files array of test files name
+ * @return list of testIds
+ */
+function extractTestIdFromFiles(files: string[]): string[] {
+  const idsArray: string[] = [];
```
Comment (FelixZilber) [inline src/commons/GitUtils.ts:54]: what happens if 2 files have the same testrails name?

File: src/commons/GitUtils.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  const testFiles: string[] = [];
+
+  for (const file of files) {
+    const fileName = path.basename(file);
+    if (isTestFile(fileName)) {
+      testFiles.push(fileName);
+    }
+  }
+  return testFiles;
+}
```
Comment (FelixZilber) [inline src/commons/GitUtils.ts:77]: let's create a function that gets (optional) folder path, and update testrails according to the test files founded in the subtree. if not folder path passed, it should work according to the current folder

File: src/commons/TestRailApi.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+    return this.instance;
+  }
+
+  /**
+   *   Updates an existing test case
+   *   @param testID test id as appear on testRail site
+   *   @param data fields data to update
+   */
+  public updateTestCase(testID: string, data: any): AxiosPromise {
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:37]: can we do a bulk update?

Comment (FelixZilber) [inline src/commons/TestRailApi.ts:37]: we don't use any

File: src/commons/TestRailApi.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    const fullUrl: string = `${BASE_URL}${ApiMethods.updateCase}${testID}`;
+    return this.makeRequest(RequestMethod.POST, fullUrl, data);
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:39]: consider rename to callApi or sendRequest

File: src/commons/TestRailService.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import { IField, TestFields, TestRailApi } from './TestRailApi';
+import { AxiosPromise } from 'axios';
+
+/**
+ * TestRail service
+ */
+export namespace TestRailService {
```
Comment (FelixZilber) [inline src/commons/TestRailService.ts:7]: rename to TestRailUtil?

File: src/commons/TestRailApi.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+   * @param reqUrl request url
+   * @param reqData json data
+   */
+  private makeRequest(reqMethod: RequestMethod, reqUrl: string, reqData?: object): AxiosPromise {
+    const requestConfig: AxiosRequestConfig = { method: reqMethod, url: reqUrl, data: reqData, auth: this.basicAuth };
+
+    try {
+      return axios(requestConfig);
+    } catch (error) {
+      throw error;
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:66]: check if promise need to be resolved here

File: src/commons/TestRailApi.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  }
+}
+
+export interface IField {
+  fieldName: string;
+  fieldOptions: any;
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:73]: not any

File: src/commons/TestRailApi.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+ */
+export namespace TestFields {
+  /**
+   * Automation Field
+   */
+  export const Automation: IField = {
+    fieldName: 'custom_automation',
+    fieldOptions: {
+      automated: 3,
+      applitoolsCandidate: 4,
```
Comment (FelixZilber) [inline src/commons/TestRailApi.ts:87]: reundant

File: src/commons/TestRailService.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  /**
+   *   Change test field
+   *   @param testID test id as appear on testRail site
+   *   @param field test field
+   *   @param option field option number
+   */
+  export function changeTestField(testID: string, field: IField, option: number): AxiosPromise {
+    const data: any = {};
```
Comment (FelixZilber) [inline src/commons/TestRailService.ts:28]: not any. looks like you need a map here

File: src/commons/TestFilesUtils.ts [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    const testFiles: Set<string> = new Set<string>();
+
+    for (const file of files) {
+      const fileName = path.basename(file);
+      if (isTestFile(fileName)) {
+        testFiles.add(fileName);
+      }
+    }
+    return testFiles;
+  }
```
Comment (FelixZilber) [inline src/commons/TestFilesUtils.ts:38]: you can use the filter function of Set for cleaner code...

File: src/commons/TestFilesUtils.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/**
+ * Return list of all files in folder and sub folder
+ * @param folderPath path of folder to get the files from
+ */
+function getListOfFilesRecursively(folderPath: string): Set<string> {
+  return getListOfFilesRecursivelyHelper(folderPath, new Set<string>());
+}
```
Comment (FelixZilber) [inline src/commons/TestFilesUtils.ts:59]: why is it needed? can getListOfFilesRecursivelyHelper get set optional instead of creating new function?

**Review by FelixZilber — APPROVED** [review]

**Pull request merged**